File: engine/julia/volatility_edge_expiration_engine.jl
# Module AI5 (Julia): Institutional Volatility Edge & Expiration Microstructure Engine
# Gravitational strike pinning physics, Greek risk budgeting, and probability of touch calculations.

module VolatilityEdgeExpirationEngine

export calculate_pinning, evaluate_vega_theta

function calculate_pinning(spot::Float64, strike::Float64, dte::Float64, oi::Int)
    dist = abs(spot - strike)
    t_factor = exp(-max(0.01, dte) * 2.0)
    pull = (oi / (dist^2 + 1.0)) * t_factor
    is_candidate = (dist < 2.0) && (dte <= 1.0) && (oi > 5000)
    return (pull = pull, is_candidate = is_candidate)
end

function evaluate_vega_theta(vega::Float64, theta::Float64, max_ratio::Float64 = 3.5)
    ratio = abs(vega) / max(1e-4, abs(theta))
    balanced = ratio <= max_ratio
    return (ratio = ratio, balanced = balanced)
end

end

File: engine/rust/src/volatility_edge_expiration_engine.rs
//! Module AI4 (Rust): Institutional Volatility Edge & Expiration Microstructure Engine.
//! High-speed SIMD strike pinning and Greek budget assessment.

#[repr(C, align(64))]
#[derive(Debug, Clone, Copy)]
pub struct VolatilityEdgeState {
    pub spot_price: f64,
    pub target_pin_strike: f64,
    pub dte_days: f64,
    pub pinning_gravitational_score: f64,
    pub portfolio_vega: f64,
    pub portfolio_theta: f64,
    pub vega_theta_ratio: f64,
    pub is_pinning_candidate: u32,
    pub risk_budget_balanced: u32,
}

impl VolatilityEdgeState {
    pub fn evaluate(spot: f64, strike: f64, dte: f64, oi: i32, vega: f64, theta: f64) -> Self {
        let dist = (spot - strike).abs();
        let t_factor = (-dte.max(0.01) * 2.0).exp();
        let pull = (oi as f64 / (dist * dist + 1.0)) * t_factor;
        let is_pin = if dist < 2.0 && dte <= 1.0 && oi > 5000 { 1 } else { 0 };

        let ratio = vega.abs() / theta.abs().max(1e-4);
        let balanced = if ratio <= 3.5 { 1 } else { 0 };

        Self {
            spot_price: spot,
            target_pin_strike: strike,
            dte_days: dte,
            pinning_gravitational_score: pull,
            portfolio_vega: vega,
            portfolio_theta: theta,
            vega_theta_ratio: ratio,
            is_pinning_candidate: is_pin,
            risk_budget_balanced: balanced,
        }
    }
}
